src/event/event_queue.rs

```diff
@@ -138,6 +162,57 @@ impl Channel {
     }
 }
 
+impl Clone for Channel {
+    /// Returns a clone of the Channel instance that shares the internal
+    /// unbounded queue with the original channel. Scheduling callbacks on the
+    /// same queue is faster than using separate channels, but might lead to
+    /// starvation if one of the threads posts significantly more callbacks on
+    /// the channel than the other one.
+    fn clone(&self) -> Self {
+        // Not referenced, we can simply clone the fields
+        if !self.has_ref {
+            return Self {
+                state: Arc::clone(&self.state),
+                has_ref: false,
+            };
+        }
+
+        let mut state = self.state.write().unwrap();
+
+        // Only need to increase the ref count since the tsfn is already referenced
+        debug_assert!(state.ref_count > 0);
+        state.ref_count += 1;
+
+        Self {
+            state: Arc::clone(&self.state),
+            has_ref: true,
+        }
+    }
+}
+
+impl Drop for Channel {
+    fn drop(&mut self) {
+        // Not a referenced event queue
+        if !self.has_ref {
+            return;
+        }
+
+        // It was only us who kept the `ChannelState` alive. No need to unref
+        // the `tsfn`, because it is going to be dropped once this function
+        // returns.
+        if Arc::strong_count(&self.state) == 1 {
```

Comment:
This is a really great trick that's much lighter weight than locking the mutex!

There is a scenario where we *could* return early that this won't catch. If there are only unreferenced clones, the `Arc::strong_count` may be `> 0`, while `ref_count == 0`. This may be a likely scenario if a user creates one unreferenced `queue` that lives a long time, but mostly uses `cx.queue()` as a one-off.

Do you think this is a likely enough scenario that it's worth the extra cost of a `read` lock on the internal state?

Reply:
If `ref_count == 0` - we'll return early at the start of the function where we check `has_ref`. I've added another optimization for the case where there are more than one references to the shared state. PTAL.